### PR TITLE
[WIP] Support Rollup External Option 

### PR DESCRIPTION
**Pull request by kristoferbaxter**
Basic tests are passing on this branch.

1. Moved to `magic-string` for mutations to source.
2. Added a new `Transform`, `ImportTransform` which removes import statements for items denoted as `external` via Rollup configuration. After Closure Compiler is finished, they are returned.
3. Switched to a default module resolution strategy of `NODE`. TBD if truly necessary.

@Andarist – might be worth trying this branch and see how it handles #11

### CODE REVIEW

**Comment by Andarist** [comment]
I've tried it on a react component lib ( https://github.com/andreypopp/react-textarea-autosize ) and got
```
[!] (closure-compiler plugin) Error: Error transforming chunk with 'closure-compiler' plugin: Google Closure Compiler exit 1: java -jar /rollup-plugin-closure-compiler/node_modules/google-closure-compiler/compiler.jar --language_out=NO_TRANSPILE --assume_function_wrapper=true --warning_level=QUIET --module_resolution=NODE --externs=/private/var/folders/ws/tlwxqvs5407f21nxy7w8bgr80000gp/T/52d79d5d-e036-4c68-b4b4-3e12273f023d --externs=/private/var/folders/ws/tlwxqvs5407f21nxy7w8bgr80000gp/T/159f75a7-34ef-4379-9608-84ab2f50ef05 --externs=/private/var/folders/ws/tlwxqvs5407f21nxy7w8bgr80000gp/T/132fd573-0a1c-4471-a788-85b61c9671ac --externs=/private/var/folders/ws/tlwxqvs5407f21nxy7w8bgr80000gp/T/2b8db9c2-ee06-4ba5-baed-4e4afa39024c --js=/private/var/folders/ws/tlwxqvs5407f21nxy7w8bgr80000gp/T/88e9bf53-edd0-4d21-91af-c6ce3da21265 --create_source_map=/private/var/folders/ws/tlwxqvs5407f21nxy7w8bgr80000gp/T/287436a9-e056-4781-acfa-87984147c7c0

/private/var/folders/ws/tlwxqvs5407f21nxy7w8bgr80000gp/T/132fd573-0a1c-4471-a788-85b61c9671ac:6: ERROR - Parse error. primary expression expected
window['default'] = default;
```

You can check out my `rollup.config.js` - for the moment I only try to bundle this:
```
  createConfig({
    output: { file: pkg.module, format: 'es' },
    browser: false
  })
```

and I've replaced uglify with `compiler({ module_resolution: 'NODE' })`. Can push out the test branch if you want.

**Comment by kristoferbaxter** [comment]
Will use your repo as a test to get more confidence in the transforms and report back.

Thanks for the suggestions!

**Comment by Andarist** [comment]
https://github.com/andreypopp/react-textarea-autosize/tree/closure-compiler-test

File: rollup.config.js [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 export default {
   input: './transpile/index.js',
-  external: ['util', 'fs', 'path', 'google-closure-compiler', 'temp-write'],
+  external: ['util', 'fs', 'path', 'google-closure-compiler', 'magic-string', 'temp-write'],
```
[the remark this answers is not in the document]
Comment (kristoferbaxter) [reply rollup.config.js]: Addressed with 81e384f.

File: src/transformers/strict.ts [hunk 2]
```diff
@@ -35,9 +36,16 @@ export default class StrictTransform extends Transform {
...
+      const source = new MagicString(code);
+
       // This will only remove the top level 'use strict' directive since we cannot
       // be certain source does not contain strings with the intended content.
-      code = code.slice(STRICT_MODE_DECLARATION_LENGTH, code.length);
+      source.remove(0, STRICT_MODE_DECLARATION_LENGTH);
+
+      return {
+        code: source.toString(),
+        map: source.toString(),
```
[the remark this answers is not in the document]
Comment (kristoferbaxter) [reply src/transformers/strict.ts]: Addressed with e6beff4.

**Comment by kristoferbaxter** [comment]
Narrowed down the problem with `react-textarea-autosize`. When it's transpiled, the export syntax is changed to a format that isn't supported yet.

```javascript
function Foo(value) {
  console.log(value);
}

export {Foo as default};
```

This is causing the current code to try and create an extern for `window['default']=default;` and a series of other problems since it believes `default` is an `ExportClosureMapping.NAMED_CONSTANT`.

Solution is a refactor of ParsingUtilities to rely more heavily on examining the AST to unwind what a variable is referencing. This means it can discover `as default` means a default export of the identifier `Foo`, and `Foo` is a function, so the result should be a `ExportClosureMapping.DEFAULT_FUNCTION`.

**Comment by kristoferbaxter** [comment]
Label was incorrectly configured when CLA bot ran.

**Pull request merged**